libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,236 @@
+         }
+      });
+   }
+
+   ~reliable_amqp_publisher_impl() {
+      stopping = true;
+
+      //drain any remaining items on queue
+      std::promise<void> shutdown_promise;
+      ctx.post([&]() {
+         shutdown_promise.set_value();
+      });
+      shutdown_promise.get_future().wait();
+
+      ctx.stop();
+      thread.join();
+
+      //if a message is in flight, let it try to run to completion
+      if(in_flight) {
+         auto until = std::chrono::steady_clock::now() + std::chrono::seconds(5);
+         ctx.restart();
+         while(in_flight && ctx.run_one_until(until));
+      }
+
+      if(message_deque.size()) {
+         try {
+            boost::filesystem::create_directories(data_file_path.parent_path());
+
+            fc::datastream<fc::cfile> file;
+            file.set_file_path(data_file_path);
+            file.open("wb");
+            fc::raw::pack(file, message_deque);
+         } FC_LOG_AND_DROP();
+      }
+   }
+
+   void retry_connection() {
+      connected = false;
+      channel.reset();
+      connection.reset();
+      handler.reset();
+      next_retry_time = std::min(next_retry_time, 30u);
+      retry_timer.expires_from_now(boost::posix_time::seconds(next_retry_time));
+      retry_timer.async_wait([&](auto ec) {
+         if(ec)
+            return;
+         handler = std::make_unique<reliable_amqp_publisher_handler>(this, ctx);
+         try {
+            connection = std::make_unique<AMQP::TcpConnection>(handler.get(), *amqp_address);
+         } catch(...) { //this should never happen, but it could technically have thrown
+            retry_connection();
+         }
+      });
+      next_retry_time *= 2;
+   }
+
+   void amqp_ready() override {
+      ilog("AMQP connection ${s} publishing to \"${e}\" established", ("s", (std::string)*amqp_address)("e", exchange));
+      next_retry_time = 1;
+      connected = true;
+      channel = std::make_unique<AMQP::TcpChannel>(connection.get());
+      channel->onError([this](const char* s) {
+         elog("Channel error for AMQP connection ${s} publishing to \"${e}\": ${m}; restarting connection", ("s", (std::string)*amqp_address)("e", exchange)("m",s));
+         retry_connection();
+      });
+   }
+
+   void amqp_error(const char* message) override {
+      wlog("AMQP connection ${a} publishing to \"${e}\" failed: ${m}; retrying", ("a", (std::string)*amqp_address)("m", message)("e", exchange));
+      retry_connection();
+   }
+
+   void pump_queue() {
+      if(stopping || in_flight || !connected || message_deque.empty())
+         return;
+
+      constexpr size_t max_msg_single_transaction = 16;
+      const unsigned msgs_this_transaction = std::min(message_deque.size(), max_msg_single_transaction);
+
+      channel->startTransaction();
+      std::for_each_n(message_deque.begin(), msgs_this_transaction, [this](const std::vector<char>& msg) {
+         channel->publish(exchange, "", msg.data(), msg.size());
```

Comment:
Seems like you should provide an option to setting a routing_key to be used. Leo's publish uses "" for exchange and a configured routing_key.

Reply:
IMO configuration of the exchange is more useful than the routing_key, but I have no qualm adding support for both.